flink-ml-iteration/src/main/java/org/apache/flink/iteration/datacache/nonkeyed/DataCache.java

```diff
@@ -0,0 +1,351 @@
+import org.apache.flink.table.runtime.util.MemorySegmentPool;
+import org.apache.flink.util.IOUtils;
+import org.apache.flink.util.function.SupplierWithException;
+
+import org.apache.commons.io.input.BoundedInputStream;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import static org.apache.flink.util.Preconditions.checkState;
+
+/** Records the data received and replays them on required. */
+@Internal
+public class DataCache<T> implements Iterable<T> {
+
+    private static final int CURRENT_VERSION = 1;
+
+    private final TypeSerializer<T> serializer;
+
+    private final FileSystem fileSystem;
+
+    private final SupplierWithException<Path, IOException> pathGenerator;
+
+    private final MemorySegmentPool segmentPool;
+
+    private final List<Segment> finishedSegments;
+
+    private SegmentWriter<T> currentWriter;
+
+    public DataCache(
+            TypeSerializer<T> serializer,
+            FileSystem fileSystem,
+            SupplierWithException<Path, IOException> pathGenerator)
+            throws IOException {
+        this(serializer, fileSystem, pathGenerator, null, Collections.emptyList());
+    }
+
+    public DataCache(
+            TypeSerializer<T> serializer,
+            FileSystem fileSystem,
+            SupplierWithException<Path, IOException> pathGenerator,
+            MemorySegmentPool segmentPool)
+            throws IOException {
+        this(serializer, fileSystem, pathGenerator, segmentPool, Collections.emptyList());
+    }
+
+    public DataCache(
+            TypeSerializer<T> serializer,
+            FileSystem fileSystem,
+            SupplierWithException<Path, IOException> pathGenerator,
+            MemorySegmentPool segmentPool,
+            List<Segment> finishedSegments)
+            throws IOException {
+        this.fileSystem = fileSystem;
+        this.pathGenerator = pathGenerator;
+        this.segmentPool = segmentPool;
+        this.serializer = serializer;
+        this.finishedSegments = new ArrayList<>();
+        this.finishedSegments.addAll(finishedSegments);
+        for (Segment segment : finishedSegments) {
+            tryCacheSegmentToMemory(segment);
+        }
+        this.currentWriter = createSegmentWriter();
+    }
+
+    public void addRecord(T record) throws IOException {
+        try {
+            currentWriter.addRecord(record);
+        } catch (SegmentNoVacancyException e) {
+            currentWriter.finish().ifPresent(finishedSegments::add);
+            currentWriter = new FileSegmentWriter<>(serializer, pathGenerator.get());
+            currentWriter.addRecord(record);
+        }
+    }
+
+    /** Finishes adding records and closes resources occupied for adding records. */
+    public void finish() throws IOException {
+        if (currentWriter == null) {
```

Comment:
Like described in the comment below, we might need to call `finish()` before reading from the data cache, which happens every time before we create a iterator. Thus finish() might be invoked multiple times. If we remove this statement, we would need to do like follows.
```java
    @Override
    public DataCacheIterator<T> iterator() {
       if (!finished){
            finish();
            finished=true;
      }
        return new DataCacheIterator<>(serializer, finishedSegments);
    }
```
Thus we would still need the `if` condition, and it might be better to directly place it inside `finish()`.